hardware_in_the_loop/software/modules/iocontroller.py

```diff
@@ -0,0 +1,167 @@
+# Extended python imports
+import serial
+import logging
+
+
+class IOController:
+    """High level python object to interface with hardware.
+
+    Used to set analog and digital pins for simulation.
+
+    https://docs.olinelectricmotorsports.com/display/AE/IO+Controller
+    """
+
+    def __init__(self, pin_info_path: str, serial_path: str):
+        # Create logger (all config should already be set by RoadkillHarness)
+        self.log = logging.getLogger(name=__name__)
+
+        self.pin_info = self.read_pin_info(path=pin_info_path)
+        try:
+            self.serial = serial.Serial(port=serial_path, baudrate=115200, timeout=5)
+        except Exception as e:
+            # Couldn't open the specified port; initialize w/o hardware for testing
+            self.log.error(f"Failed to connect to hardware at {pin_info_path}")
+            self.log.error(e)
+            self.serial = None
+
+    def set_state(self, pin: str, value) -> None:
+        """Set the value of an IO pin in the HitL system
+
+        Args:
+            pin (str): The name of the pin to update (e.x. THROTTLE_PEDAL_1)
+
+            value (int or float): The value to set the pin to (e.x. 2.5)
+                - 0 or 1 for digital, volts for analog
+
+        Message format:
+            2 bytes.
+            Byte 1:
+                Bit 0: 1 (indicates a set request)
+                Bits 1-7: Address (int)
+
+            Byte 2: Value
+
+            Value format
+                Digital: 0 or 1 (easy enough)
+                Analog: <5 bit>.<3 bit> floating point value (0 to 31.875 V)
+                    For example 5V is 00101000 or 0x40
+                    For example, 2.5V is 00010100 or 0x20
+                    To convert from int, multiply by 8 (or bitshift left 3 times)
+        """
+        # If no hardware, log an error
+        if not self.serial:
+            self.log.error("Could not set state; no hardware connection")
+            return
+
+        address = self.pin_info[pin]["address"] + 128  # set the leftmost bit to 1 to designate message as setter
```

Comment:
```python
address |= 1 << 7
```

for bitwise math, lets use bitwise operators to be more clear whats going on

Reply:
sure sounds good.